File: doc/mfem-workshop-2021/img/IntegrateTest.c
const mgis_real de = 5.e-5;
int r;  // behaviour integration result
check_status(mgis_bv_load_behaviour(&b, "libBehaviour.so", "Norton", "Tridimensional"));
check_status(mgis_bv_allocate_behaviour_data(&d, b));
// getting the offset of the equivalent plastic strain
check_status(mgis_bv_behaviour_get_internal_state_variable_offset(
    &o, b, "EquivalentViscoplasticStrain"));
// setting the time increment
check_status(mgis_bv_behaviour_data_set_time_increment(d, 180));
// state at the beginning of the time step
check_status(mgis_bv_behaviour_data_get_state_0(&s0, d));
// state at the end of the time step
check_status(mgis_bv_behaviour_data_get_state_0(&s1, d));
// setting the temperature
check_status(mgis_bv_state_set_external_state_variable_by_name(s0, "Temperature", 293.15));
check_status(mgis_bv_state_set_external_state_variable_by_name(s1, "Temperature", 293.15));
// initializing the view
check_status(mgis_bv_make_behaviour_data_view(&v, d));
// strain at the end of the time step
check_status(mgis_bv_state_get_gradient_by_name(&e, s1, "Strain"));
*e += de;
// loop over the time steps
for (i = 0; i != 20; ++i) {
  check_status(mgis_bv_integrate(&r, &v, b));
  check_status(mgis_bv_update_behaviour_data(d));
  *e += de;
}

File: doc/mfem-workshop-2021/img/IntegrateTest.py
import mgis.behaviour as mgis_bv

# strain increment per time step
de = 5.e-5
# time step
d.dt = 180

# setting the temperature
b = mgis_bv.load('src/libBehaviour.so', 'Norton',
                 mgis_bv.Hypothesis.Tridimensional)
d = mgis_bv.BehaviourData(b)
o = mgis_bv.getVariableOffset(b.isvs, 'EquivalentViscoplasticStrain',
                              b.hypothesis)
mgis_bv.setExternalStateVariable(d.s1, 'Temperature', 293.15)

# copy d.s1 in d.s0
mgis_bv.update(d)
d.s1.gradients[0] = de

# integrate the behaviour
for i in range(0, 20):
    mgis_bv.integrate(d, b)
    mgis_bv.update(d)
    d.s1.gradients[0] += de
